README.md
# NAME

Dist::Iller - A Dist::Zilla & Pod::Weaver preprocessor

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.14+-blue.svg" alt="Requires Perl 5.14+" />
    <img src="https://img.shields.io/badge/coverage-84.4%25-orange.svg" alt="coverage 84.4%" />
    <a href="https://github.com/Csson/p5-Dist-Iller/actions?query=workflow%3Amakefile-test"><img src="https://img.shields.io/github/workflow/status/Csson/p5-Dist-Iller/makefile-test" alt="Build status at Github" /></a>
    </p>
</div>

# VERSION

Version 0.1411, released 2020-01-01.

# SYNOPSIS

    # dzil new, but...
    $ dzil new -P DistIller::AMintingProvider My::Module

    $ cd My/Module

    # ...all other commands can be used via iller
    $ iller build

# STATUS

This is alpha software. Anything can change at any time.

It is mostly here to document how I build my distributions. It is perfectly fine to use `dzil` with a distribution built with `Dist::Iller` (after a fork, for example).

# DESCRIPTION

Dist::Iller is a [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla) and [Pod::Weaver](https://metacpan.org/pod/Pod::Weaver) preprocessor. It comes with a command line tool (`iller`) which is a `dzil` wrapper: When run, it first generates
files specified in `iller.yaml` in the current directory and then executes `dzil` automatically. (Since `iller` requires that an `iller.yaml` is present, `iller new ...` does not work.)

The `doctype` key in a document in `iller.yaml` matches a camelized class in the `Dist::Iller::DocType` namespace; so `doctype: dist` is parsed by [Dist::Iller::DocType::Dist](https://metacpan.org/pod/Dist::Iller::DocType::Dist).

## iller.yaml

This is the general syntax of an `iller.yaml` file:

    ---
    # This specifies that this yaml document will generate dist.ini.
    doctype: dist

    # This generates the top part of C<dist.ini>. C<author> can be a list or string.
    header:
      name: My-Module
      author: Ex Ample <ample@example.org>
      license: Perl_5
      copyright_holder: Ex Ample
      copyright_year: 2015

    # It is possible to list all prereqs. The groups are specified in CPAN::Meta::Spec.
    # Minimum version numbers are optional.
    prereqs:
      runtime:
        requires:
          - perl: 5.010001
          - Moose

    # List all plugins under the 'plugins' key.
    # Each +plugin item is a Dist::Zilla> plugin.
    # All commands for Dist::Iller is prepended with a +.
    plugins:
      # Includes all plugins specified in Dist::Iller::Config::My::Config
      - +config: My::Config

      - +plugin: DistIller::MetaGeneratedBy
      - +plugin: AutoVersion
      - +plugin: GatherDir

      # 'dir' is a parameter for ShareDir
      - +plugin: ShareDir
        dir: myshare

    [...]

    ---
    # Here starts the weaver.ini configuration.
    doctype: weaver

    plugins:
      # Same Dist::Iller::Config as in the 'dist' document
      - +config: My::Config

      # Use PluginBundles
      - +plugin: '@CorePrep'

      - +plugin: -SingleEncoding

      - +plugin: Name

      - +plugin: Version
        format: Version %v, released %{YYYY-MM-dd}d.

      - +plugin: prelude
        +base:  Region

      - +plugin: List
        +base: -Transformer
        +in: Elemental
        transformer: List

    [...]

    ---
    # Here starts the .gitignore configuration
    doctype: gitignore

    always:
      - /.build
      - /_build*
      - /Build
      - MYMETA.*
      - '!META.json'
      - /.prove
    ---
    # No configuration for .cpanfile, but by having a YAML document for it, it gets generated from
    # the prereqs listed in the 'dist' document
    doctype: cpanfile

## Rationale

PluginBundles for both [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla) and [Pod::Weaver](https://metacpan.org/pod/Pod::Weaver) have a few downsides:

- Mixes code and configuration.
- Not straightforward to remove or replace specific plugins for a certain distribution
- Difficult to insert a plugin before another plugin for a certain distribution.
- PluginBundles can change after a distribution has been released.
- Difficult for others to understand/know which plugins actually were in effect when the distribution was built.

`Dist::Iller` tries to solve this:

- Dist::Iller configs (similar to PluginBundles) has their own `iller.yaml` (normally in `share/`) where plugins are specified. See tests and [Dist::Iller::Config::Author::CSSON](https://metacpan.org/pod/Dist::Iller::Config::Author::CSSON).
- Since `dist.ini` and `weaver.ini` are generated each time `iller` is run, the plugins listed in them are those that were used to build the distribution.
- Remove a plugin:

      - +remove_plugin: GatherDir

- Insert a plugin:

      - +add_plugin: Git::GatherDir
        +before: AutoVersion

- Replace a plugin:

      - +replace_plugin: ShareDir
        +with: ShareDir::Tarball

- Set more attributes for an already included plugin:

      - +extend_plugin: Git::GatherDir
        exclude_match:
          - examples/.*\.html

# SEE ALSO

- [Dist::Zilla](https://metacpan.org/pod/Dist::Zilla)
- [Pod::Weaver](https://metacpan.org/pod/Pod::Weaver)
- [Dist::Iller::Config::Author::CSSON](https://metacpan.org/pod/Dist::Iller::Config::Author::CSSON)

# SOURCE

[https://github.com/Csson/p5-Dist-Iller](https://github.com/Csson/p5-Dist-Iller)

# HOMEPAGE

[https://metacpan.org/release/Dist-Iller](https://metacpan.org/release/Dist-Iller)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2021 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Dist-Iller

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.1411 2021-01-01T19:00:59Z UTC
 [Stats & Dependency Changes]
 - 17 files changed, 18 insertions(+), 18 deletions(-)
 - (dev sug) ~ Dist::Iller 0.1409 --> 0.1410

 [Test Changes]
 - Fix a test failure that started to appear this year, caused by
   yesterday's change regarding copyright_year. Fixes #5, thanks SREZIC!

0.1410 2020-12-31T17:59:13Z UTC
 [Enhancements]
 - copyright_year in dist now defaults to the current year. Previously, the
   config needed to be updated each year.

 [Requirements]
 - Required Perl is 5.14, since that is what Dist::Zilla requires.

 [Stats & Dependency Changes]
 - 19 files changed, 47 insertions(+), 37 deletions(-)
 - (run req) ~ Dist::Zilla 5.000 --> 6.015
 - (run req) ~ Moose 2.1400 --> 2.1803
 - (run req) ~ Types::Standard 1.000000 --> 1.000006
 - (run req) ~ namespace::autoclean 0.18 --> 0.22
 - (run req) ~ perl 5.010000 --> 5.014000
 - (dev req) + Badge::Depot::Plugin::Githubactions (any)
 - (dev req) - Badge::Depot::Plugin::Cpantesters
 - (dev req) - Badge::Depot::Plugin::Kwalitee
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0327 --> 0.0328

0.1409 2020-12-27T17:05:17Z UTC
 [New Features]
 - Added new DocType: Global, used for settings common to multiple DocTypes
   (so far only distribution_name)
 - The global document (if it exists in iller.yaml) will be passed to
   Config objects
 - in DocType/Gitignore, it is now possible to use $self.distribution_name
   in 'always' and 'onexist', so that distribution build directories are
   automatically ignored.

 [Stats & Dependency Changes]
 - 21 files changed, 213 insertions(+), 53 deletions(-)
 - (run req) ~ List::Util 1.42 --> 1.45
 - (dev req) +
   Dist::Zilla::Plugin::Author::CSSON::GithubActions::Workflow::TestWithMakefile
   0.0106
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale 0.057
 - (dev req) ~ Dist::Zilla 5.000 --> 6.015
 - (dev req) ~ Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.007 --> 0.009
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0100
   --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent 0.008 --> 0.011
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild 0.151680 -->
   0.170880
 - (dev req) ~ Dist::Zilla::Plugin::Git 2.036 --> 2.046
 - (dev req) ~ Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.013 -->
   0.014
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors 0.017 --> 0.035
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package 2.003001 -->
   2.004003
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod 0.150250 --> 0.163250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests 0.028 --> 0.029
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile 2.054 --> 2.058
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs 0.021 --> 0.027
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version 1.05 --> 1.09
 - (dev req) - Badge::Depot::Plugin::Travis
 - (dev req) - Dist::Zilla::Plugin::Test::Kwalitee::Extra
 - (dev req) - Dist::Zilla::Plugin::TravisYML
 - (dev sug) ~ Dist::Iller 0.1408 --> 0.1409
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0309 --> 0.0327

0.1408 2016-03-12T20:16:19Z UTC
 [New Features]
 - Can add all prereqs as authordeps (add_prereqs_as_authordeps)

 [Stats & Dependency Changes]
 - 17 files changed, 37 insertions(+), 17 deletions(-)
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) ~ Dist::Iller 0.1407 --> 0.1408
 - (dev sug) ~ Dist::Iller::Config::Author::CSSON 0.0308 --> 0.0309

0.1407 2016-02-18T19:06:10Z UTC
 [New Features]
 - Configs can now add a list of default prereq versions. If no version is
   given in iller.yaml, the default version is used. Only already added
   prereqs are checked for default versions.
 - Add Gitignore doctype.

 [Stats & Dependency Changes]
 - 20 files changed, 186 insertions(+), 22 deletions(-)
 - (run req) ~ Carp (any) --> 1.38
 - (run req) ~ DateTime (any) --> 1.05
 - (run req) ~ File::ShareDir (any) --> 1.102
 - (run req) ~ File::chdir (any) --> 0.1010
 - (run req) ~ JSON::MaybeXS (any) --> 1.003003
 - (run req) ~ List::MoreUtils (any) --> 0.408
 - (run req) ~ List::Util (any) --> 1.42
 - (run req) ~ Module::Load (any) --> 0.26
 - (run req) ~ MooseX::AttributeDocumented (any) --> 0.1003
 - (run req) ~ MooseX::AttributeShortcuts (any) --> 0.028
 - (run req) ~ MooseX::StrictConstructor (any) --> 0.19
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ PerlX::Maybe (any) --> 1.001
 - (run req) ~ Safe::Isa (any) --> 1.000005
 - (run req) ~ Try::Tiny (any) --> 0.24
 - (run req) ~ Types::Path::Tiny (any) --> 0.005
 - (run req) ~ YAML::Tiny (any) --> 1.67
 - (run req) ~ version (any) --> 0.9913
 - (test req) ~ File::chdir (any) --> 0.1010
 - (dev req) + Dist::Zilla 5.000
 - (dev req) + Dist::Zilla::Plugin::Git 2.036
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Badge::Depot (any) --> 0.0103
 - (dev req) ~ Dist::Zilla::Plugin::ChangeStats::Dependencies::Git (any)
   --> 0.0100
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.151680
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.017
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.003001
 - (dev req) ~ Dist::Zilla::Plugin::NextRelease::Grouped (any) --> 0.0200
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006

Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.031.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "(deprecated)",
  "dist_author" => [
    "Erik Carlsson <info\@code301.com>"
  ],
  "dist_name" => "Dist-Iller",
  "dist_version" => "0.1102",
  "license" => "perl",
  "module_name" => "Dist::Iller",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Config::INI" => 0,
    "Dist::Zilla" => "5.000",
    "Dist::Zilla::Plugin::CheckChangesHasContent" => 0,
    "Dist::Zilla::Plugin::ConfirmRelease" => 0,
    "Dist::Zilla::Plugin::ContributorsFromGit" => 0,
    "Dist::Zilla::Plugin::CopyFilesFromBuild" => 0,
    "Dist::Zilla::Plugin::ExtraTests" => 0,
    "Dist::Zilla::Plugin::Git::Check" => 0,
    "Dist::Zilla::Plugin::Git::Commit" => 0,
    "Dist::Zilla::Plugin::Git::GatherDir" => 0,
    "Dist::Zilla::Plugin::Git::Push" => 0,
    "Dist::Zilla::Plugin::Git::Tag" => 0,
    "Dist::Zilla::Plugin::GithubMeta" => 0,
    "Dist::Zilla::Plugin::License" => 0,
    "Dist::Zilla::Plugin::Manifest" => 0,
    "Dist::Zilla::Plugin::ManifestSkip" => 0,
    "Dist::Zilla::Plugin::MetaJSON" => 0,
    "Dist::Zilla::Plugin::MetaNoIndex" => 0,
    "Dist::Zilla::Plugin::MetaYAML" => 0,
    "Dist::Zilla::Plugin::ModuleBuildTiny" => 0,
    "Dist::Zilla::Plugin::NextRelease" => 0,
    "Dist::Zilla::Plugin::NextVersion::Semantic" => 0,
    "Dist::Zilla::Plugin::OurPkgVersion" => 0,
    "Dist::Zilla::Plugin::PodSyntaxTests" => 0,
    "Dist::Zilla::Plugin::PodWeaver" => "4.000",
    "Dist::Zilla::Plugin::PodnameFromFilename" => 0,
    "Dist::Zilla::Plugin::Prereqs::FromCPANfile" => 0,
    "Dist::Zilla::Plugin::PreviousVersion::Changelog" => 0,
    "Dist::Zilla::Plugin::ReadmeAnyFromPod" => 0,
    "Dist::Zilla::Plugin::ReversionOnRelease" => 0,
    "Dist::Zilla::Plugin::ShareDir" => 0,
    "Dist::Zilla::Plugin::Test::EOL" => 0,
    "Dist::Zilla::Plugin::TestRelease" => 0,
    "Dist::Zilla::Plugin::UploadToCPAN" => 0,
    "Dist::Zilla::Plugin::UploadToStratopan" => 0,
    "Dist::Zilla::Role::PluginBundle::Config::Slicer" => 0,
    "Dist::Zilla::Role::PluginBundle::Easy" => 0,
    "Dist::Zilla::Role::PluginBundle::PluginRemover" => 0,
    "File::chdir" => 0,
    "Git::Wrapper" => 0,
    "IPC::Run" => 0,
    "List::AllUtils" => 0,
    "Moose" => "2.0000",
    "MooseX::AttributeShortcuts" => 0,
    "Path::Tiny" => 0,
    "Pod::Weaver::Config::Assembler" => 0,
    "Types::Standard" => 0,
    "namespace::autoclean" => 0,
    "perl" => "5.010001"
  },
  "script_files" => [
    "script/iller"
  ],
  "share_dir" => {
    "module" => {
      "Dist::Zilla::MintingProfile::Iller" => "profiles"
    }
  },
  "test_requires" => {
    "Test::More" => "0.96",
    "Test::NoTabs" => 0
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.3601",
  "Test::More" => "0.96",
  "Test::NoTabs" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Pod/Weaver/PluginBundle/Iller.pm
use 5.10.1;

package Pod::Weaver::PluginBundle::Iller;

# VERSION

use strict;
use warnings;
use Pod::Weaver::Config::Assembler;
use Path::Tiny;

sub xp {
    Pod::Weaver::Config::Assembler->expand_package(shift);
}

sub mvp_bundle_config {
    my @plugins = ();

    # check git config
    my $include_default_github = 0;
    my $git_config = path('.git/config');
    if($git_config->exists) {
        my $git_config_contents = $git_config->slurp_utf8;
        if($git_config_contents =~ m{github\.com:([^/]+)/(.+)\.git}) {
            $include_default_github = 1;
        }
        else {
            warn ('[PW/@Iller] No github url found');
        }
    }

    push @plugins => (
        ['@Iller/CorePrep',       xp('@CorePrep'),       { } ],
        ['@Iller/SingleEncoding', xp('-SingleEncoding'), { } ],
        ['@Iller/Name',           xp('Name'),            { } ],
        ['@Iller/Version',        xp('Version'),         { format => q{Version %v, released %{YYYY-MM-dd}d.} } ],
        ['@Iller/Prelude',        xp('Region'),          { region_name => 'prelude' } ],
    );

    foreach my $plugin (qw/Synopsis Description Overview Stability/) {
        push @plugins => ['@Iller/'.$plugin, xp('Generic'), { header => uc $plugin } ];
    }

    foreach my $plugin ( ['Attributes', 'attr'],
                         ['Methods', 'method'],
                         ['Functions', 'func'],
    ) {
        push @plugins => [ $plugin->[0], xp('Collect'), { command => $plugin->[1], header => uc $plugin->[0] } ];
    }
    push @plugins => (
        ['@Iller/Leftovers',             xp('Leftovers'), { } ],
        ['@Iller/postlude',              xp('Region'),    { } ],
        (
            !$ENV{'ILLER_MINTING'} && $include_default_github ?
            ['@Iller/Source::DefaultGitHub', xp('Source::DefaultGitHub'), { text => 'L<%s>' } ]
            :
            ()
        ),
        ['@Iller/Homepage::DefaultCPAN', xp('Homepage::DefaultCPAN'), { text => 'L<%s>' } ],
        ['@Iller/Authors',               xp('Authors'),   { } ],
        ['@Iller/Legal',                 xp('Legal'),     { } ],

        ['@Iller/List', xp('-Transformer'), { transformer => 'List' } ],
    );


    return @plugins;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share module => "Dist::Zilla::MintingProfile::DistIller::Basic", "profiles";


my %WriteMakefileArgs = (
  "ABSTRACT" => "A Dist::Zilla & Pod::Weaver preprocessor",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Dist-Iller",
  "EXE_FILES" => [
    "script/iller"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "Dist::Iller",
  "PREREQ_PM" => {
    "Carp" => "1.38",
    "DateTime" => "1.05",
    "Dist::Zilla" => "6.015",
    "File::ShareDir" => "1.102",
    "File::chdir" => "0.1010",
    "JSON::MaybeXS" => "1.003003",
    "List::MoreUtils" => "0.408",
    "List::Util" => "1.45",
    "Module::Load" => "0.26",
    "Moose" => "2.1803",
    "MooseX::AttributeDocumented" => "0.1003",
    "MooseX::AttributeShortcuts" => "0.028",
    "MooseX::StrictConstructor" => "0.19",
    "Path::Tiny" => "0.072",
    "PerlX::Maybe" => "1.001",
    "Safe::Isa" => "1.000005",
    "String::CamelCase" => "0.02",
    "Text::Diff" => "1.43",
    "Try::Tiny" => "0.24",
    "Types::Path::Tiny" => "0.005",
    "Types::Standard" => "1.000006",
    "YAML::Tiny" => "1.67",
    "namespace::autoclean" => "0.22",
    "version" => "0.9913"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::chdir" => "0.1010",
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Syntax::Feature::Qi" => 0,
    "Syntax::Feature::Qs" => 0,
    "Test::Differences" => 0,
    "Test::More" => "0.96",
    "syntax" => 0
  },
  "VERSION" => "0.1412",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => "1.38",
  "DateTime" => "1.05",
  "Dist::Zilla" => "6.015",
  "ExtUtils::MakeMaker" => 0,
  "File::ShareDir" => "1.102",
  "File::Spec" => 0,
  "File::chdir" => "0.1010",
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeXS" => "1.003003",
  "List::MoreUtils" => "0.408",
  "List::Util" => "1.45",
  "Module::Load" => "0.26",
  "Moose" => "2.1803",
  "MooseX::AttributeDocumented" => "0.1003",
  "MooseX::AttributeShortcuts" => "0.028",
  "MooseX::StrictConstructor" => "0.19",
  "Path::Tiny" => "0.072",
  "PerlX::Maybe" => "1.001",
  "Safe::Isa" => "1.000005",
  "String::CamelCase" => "0.02",
  "Syntax::Feature::Qi" => 0,
  "Syntax::Feature::Qs" => 0,
  "Test::Differences" => 0,
  "Test::More" => "0.96",
  "Text::Diff" => "1.43",
  "Try::Tiny" => "0.24",
  "Types::Path::Tiny" => "0.005",
  "Types::Standard" => "1.000006",
  "YAML::Tiny" => "1.67",
  "namespace::autoclean" => "0.22",
  "syntax" => 0,
  "version" => "0.9913"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}
